File: benchmarks/micro/NullBranchPruning.c
#include <stdint.h>
#include <unistd.h>

__attribute__((noinline)) static int32_t score(int64_t *ptr, int64_t seed) {
    if (ptr != 0) {
        if (ptr == 0) {
            return 1;
        }

        return (int32_t)((seed & 7) + 2);
    }

    return 0;
}

int main(void) {
    int64_t total = (int64_t)getpid();
    int64_t slot = total;
    int64_t *ptr = &slot;

    for (int32_t i = 0; i < 200000; i += 1) {
        total += score(ptr, total + (int64_t)i);
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/NullBranchPruning.rs
#[inline(never)]
fn score(ptr: *mut i64, seed: i64) -> i32 {
    if !ptr.is_null() {
        if ptr.is_null() {
            return 1;
        }

        ((seed & 7) + 2) as i32
    } else {
        0
    }
}

fn main() {
    let mut total = std::process::id() as i64;
    let mut slot = total;
    let ptr = &mut slot as *mut i64;

    for i in 0_i32..200_000 {
        total += score(ptr, total + i as i64) as i64;
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
